localgov_events.module

```diff
@@ -0,0 +1,169 @@
+<?php
+/**
+ * @file
+ * LocalGov Events module file.
+ */
+
+use Drupal\Core\Entity\Display\EntityFormDisplayInterface;
+use Drupal\Core\Entity\Display\EntityViewDisplayInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\localgov_events\Form\EventsAddEditCallbacks;
+
+/**
+ * Implements hook_modules_installed().
+ */
+function localgov_events_modules_installed($modules) {
+  // Configure optional fields.
+  $directory_page = in_array('localgov_directories_page', $modules);
+  $directory_venue = in_array('localgov_directories_venue', $modules);
+  if ($directory_page || $directory_venue) {
+    \Drupal::service('config.installer')->installOptionalConfig();
```

Comment:
I wonder why we need to explicitly install the optional config files given that Drupal does it anyway at module install time.  More a question for Ekes really since he wrote a [similar piece of code](https://github.com/localgovdrupal/localgov_directories/blob/2b24950cc6e9c6f45829ef7fdc66ec0c6197407f/localgov_directories.module#L49) in the localgov_directories module.  Ekes, any idea?

Also, this [hook_modules_installed()](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_modules_installed/9.0.x) has a second parameter called $is_syncing which, when true, tells us not to update configs.

FAO @ekes

Reply:
So the reason we've been re-triggering optional configuration scanning has been the optional fields/modules. For example:

If you install services and step-by-step (either services first, then step-by-step, or both at the same time) you will get a field to post step-by-step into a service.
If you install step-by-step without services you don't get the field to post your step-by-step in services - obviously it makes no sense, and the dependency isn't there. If you then _afterwards_ install services, Drupal doesn't rescan the optional configuration of step-by-step to find the optional field dependent on services. But we would like the field to be added if someone opts into having services at a later stage. So we trigger rescanning the configuration to add the field.

Reply:
I can confirm the `\Drupal::service('config.installer')->installOptionalConfig();` is needed. There is a test that checks if the venue field exists if the localgov_directories_venue module is enabled after `locagov_events`, this fails without this code.

Reply:
> If you then afterwards install services, Drupal doesn't rescan the optional configuration of step-by-step to find the optional field dependent on services.

That is really weird.  I thought Drupal would rescan all the optional configs up on a new module install.  Steve has [mentioned](https://github.com/localgovdrupal/localgov_events/blob/bf973e716088f6d0eaa830be6f20df501dcbb7da/localgov_events.install#L34) a similar bug report that Finn has raised with the [Simple sitemap project](https://www.drupal.org/project/simple_sitemap/issues/3156080).  Are these similar issues?

The [documentation is saying](https://www.drupal.org/docs/drupal-apis/configuration-api/configuration-storage#s-optional-configuration-for-an-extension) "For example, in the scenario that module A has optional configuration which requires module B, but module A is installed first and module B some time later, then module A's config/optional directory will be scanned at that time for newly met dependencies, and the configuration will be installed then".  If I believe this, then config/optional/field.field.node.localgov_event.localgov_event_venue.yml should be picked up when localgov_directiry_venu is installed.

On the other hand, the [change record is saying](https://www.drupal.org/node/2453919) "When a module is installed that provides a configuration entity type, for example Views, all installed modules' config/optional directories are searched for any configuration entities of the new type."  If I believe this, then when the field module and the localgov_event module are already installed, no optional **field_config** config entity file (e.g. field.field.node.localgov_event.localgov_event_venue.yml) would be scanned when the localgov_directory_venue module is installed.  This version matches with what Ekes is saying.